engine/gamesys/src/gamesys/components/comp_model.cpp

```diff
@@ -280,46 +280,56 @@ namespace dmGameSystem
-            texture = material_info->m_Textures[texture_unit].m_Texture;
-        if (texture)
-            return texture;
-
-        // The textures which are set in the .material file
-        material = material_info->m_Material;
-        if (material)
-            texture = material->m_Textures[texture_unit];
-        if (texture)
-            return texture;
+        // If the model has textures, let's find the one associated with the sampler name
+        for (uint32_t i = 0; i < material_info->m_TexturesCount; ++i)
+        {
+            if (material_info->m_Textures[i].m_SamplerNameHash == sampler_name_hash)
+                return material_info->m_Textures[i].m_Texture;
+        }
+
+        // If it is contains materials, use that
+        if (material_texture_index < material->m_NumTextures)
+            return material->m_Textures[material_texture_index];
 
         return 0;
     }
 
-    dmGraphics::HTexture GetMaterialTexture(const ModelComponent* component, uint32_t material_index, uint32_t texture_unit)
+    // A bit legacy, as we don't want to set textures based on index anymore, but sampler names.
+    static TextureResource* GetTextureResource(const ModelComponent* component, uint32_t material_index, uint32_t material_texture_index)
     {
-        TextureResource* texture = GetTextureResource(component, material_index, texture_unit);
-        return texture ? texture->m_Texture : 0;
+        MaterialResource* material = GetMaterialResource(component, component->m_Resource, material_index);
+        MaterialInfo* material_info = &component->m_Resource->m_Materials[material_index];
+
+        TextureResource* texture_res = component->m_Textures[material_texture_index]; // Check if it's overridden
+        if (!texture_res && material_texture_index < material_info->m_TexturesCount)
+        {
+            texture_res = material_info->m_Textures[material_texture_index].m_Texture; // Check if the model has a texture
+        }
+        if (!texture_res && material_texture_index < material->m_NumTextures)
+        {
+            texture_res = material->m_Textures[material_texture_index]; // Check if the material has a texture
+        }
+        return texture_res;
     }
 
+    static void FillTextures(dmRender::RenderObject* ro, const ModelComponent* component, uint32_t material_index)
+    {
+        MaterialResource* material = GetMaterialResource(component, component->m_Resource, material_index);
+        for(uint32_t i = 0; i < material->m_NumTextures; ++i)
+        {
+            TextureResource* texture_res = component->m_Textures[i];
+            if (!texture_res)
+            {
+                texture_res = GetTextureFromSamplerNameHash(&component->m_Resource->m_Materials[material_index], material, i, material->m_SamplerNames[i]);
+            }
+
+            ro->m_Textures[i] = texture_res ? texture_res->m_Texture : 0;
+        }
+    }
```

Comment:
Perhaps another candidate for GameSys generalization later on, so we can more easily use the same functionality for all components.